<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных-для-модели" data-toc-modified-id="Подготовка-данных-для-модели-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных для модели</a></span></li><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Выделение-характеристик-и-целевого-признака" data-toc-modified-id="Выделение-характеристик-и-целевого-признака-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Выделение характеристик и целевого признака</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Создание-матриц-шифрования-и-дешифрования" data-toc-modified-id="Создание-матриц-шифрования-и-дешифрования-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Создание матриц шифрования и дешифрования</a></span></li><li><span><a href="#Шифрование-данных" data-toc-modified-id="Шифрование-данных-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Шифрование данных</a></span></li><li><span><a href="#Подготовка-данных-для-обучения-модели" data-toc-modified-id="Подготовка-данных-для-обучения-модели-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Подготовка данных для обучения модели</a></span></li><li><span><a href="#Проверка-на-модели-линейной-регрессии" data-toc-modified-id="Проверка-на-модели-линейной-регрессии-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Проверка на модели линейной регрессии</a></span></li><li><span><a href="#Проверка-на-модели-из-Scikit-Learn" data-toc-modified-id="Проверка-на-модели-из-Scikit-Learn-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Проверка на модели из Scikit Learn</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

# Загрузка библиотек

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
# определим константы
RANDOM_STATE = 12345
np.random.seed(42)

## Подготовка данных для модели

## Загрузка данных

### Загрузка данных

In [3]:
def first_look(df, num_of_srtings=5):
    print('Общая информация')
    display(df.info())
    
    print(f'Первые {num_of_srtings} строк(и) данных')
    display(df.head(num_of_srtings))
    
    print('Основные статистические характеристики данных')
    display(df.describe())
    print('Количество пропусков:')
    print(df.isna().sum())
    print()
    
    print('Количество дубликатов:', df.duplicated().sum())

In [4]:
data = pd.read_csv('/datasets/insurance.csv')

In [5]:
first_look(data)

Общая информация
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


None

Первые 5 строк(и) данных


Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

Основные статистические характеристики данных


Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

Количество пропусков:
Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

Количество дубликатов: 153


In [6]:
data[data.duplicated()].head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
281    1     39.0   48100.0            1                  0
488    1     24.0   32900.0            1                  0
513    0     31.0   37400.0            2                  0
718    1     22.0   32600.0            1                  0
785    0     20.0   35800.0            0                  0

Дубликаты не несут полезной для обучения модели информации, поэтому удалим их.

In [7]:
data = data.drop_duplicates()

In [8]:
print('Количество дубликатов:', data.duplicated().sum())

Количество дубликатов: 0


### Выделение характеристик и целевого признака

Разобьем имеющиеся данные на характеристики и целевой признак.

In [9]:
features = data.drop('Страховые выплаты', axis=1)
target = data['Страховые выплаты']

 Проверим верность разбиения

In [10]:
# размерности характеристик и целевого признака
features.shape, target.shape

((4847, 4), (4847,))

### Вывод

На этапе подготовки данных мы загрузили данные из файла '/datasets/insurance.csv'. Нам доступны данные 5000 клиентов. По каждому клиенту предоставлены значения 4х числовых характеристик.

Пропуски отсутствуют. В данных найдены 153 дубликата. Дубликаты удалены.

Данные разбиты на характеристики (пол, возраст, зарплата застрахованного и количество членов семьи) и целевой признак – количество страховых выплат.


## Умножение матриц

Для того, чтобы зашифровать данные клиентов, умножим матрицу признаков на обратимую квадратную матрицу, размерность которой совпадает с числом характеристик. Докажем, что такое умножение не изменит качества линейной регрессии.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Утверждение 1**. Качество модели линейной регрессии не изменится при умножении матрицы характеристик на обратимую матрицу, имеющую размерность, равную количеству характеристик, описывающих данные.

**Доказательство**. Для нахождения вектора весов $w$ используется нормальное уравнение, которое имеет вид:

$$
        X^T X w = X^T y\label{eq1}\tag{1}   
$$              

Обозначим через $Z$ матрицу, которую получаем при умножении матрицы  характеристик $X$ на обратимую матрицу $B$:

$$
Z = X B\label{eq2}\tag{2}
$$

Составим для $Z$ уравнение, аналогичное (1) и посмотрим, как его решение $\hat w$ связано с решением $w$ уравнения (1)

$$
        Z^T Z \hat w = Z^T y\label{eq3}\tag{3}   
$$  

Подставим в выражение (3) вместо $Z$ правую сторону выражения (2):

$$
        (X B)^T (X B) \hat w = (X B)^T y\label{eq4}\tag{4}  
$$

Справедлива следующая теорема: транспонирование произведения матриц равно произведению транспонированных матриц, взятых в обратном порядке, то есть

$$
(X B)^T = B^T X^T
$$

Воспользуемся ею, чтобы раскрыть скобки в выражении (4).

$$
        B^T X^T X B \hat w = B^T X^T y  
$$

Домножим обе части равенства на $(B^T)^{-1}$ слева:

$$
        (B^T)^{-1} B^T X^T X B \hat w = (B^T)^{-1} B^T X^T y
$$

$$
        X^T X B \hat w = X^T y\label{eq5}\tag{5}  
$$

Сравним выражение (5) с выражением (1): $X^T X w = X^T y$

Оба выражения имеют одинаковую структуру. Фактически, выражение (5) можно получить из (1) заменив $w$ на $B \hat w$.

Таким образом, мы получили соотношения, которые связывают вектора весов $w$ и $\hat w$ линейной регрессии до преобразования признаков с помощью матрицы $B$ и после:

$$
w = B \hat w\label{eq6}\tag{6}
$$

$$
\hat w = B^{-1}  w\label{eq7}\tag{7}
$$

Выясним, как отреагируют на изменение вектора весов предсказания линейной регрессии. Вектор предсказаний рассчитывается по формуле:

$$
a = X w
$$

Посмотрим на вектор предсказаний $\hat a = Z \hat w$, который получается после шифрования данных пользователей по формуле (2). Подставим в это выражение $X B$ вместо $Z$ и $B^{-1}  w$ вместо $\hat w$:

$$
\hat a = Z \hat w \stackrel{(2)}{=} X B\hat w \stackrel{(7)}{=} X B B^{-1} w = X E w = X w= a
$$

Таким образом видим, что предсказания линейной регрессии никак не изменятся при умножении матрицы характеристик на обратимую матрицу. А следовательно, не изменится и качество модели. Что и требовалось доказать.

Выше приведено строгое доказательство утверждения 1 с помощью линейной алгебры. Его можно и более наглядно проиллюстрировать, если вспомнить геометрический смысл умножения на квадратную матрицу. Существует взаимно однозначное соответствие между квадратными матрицами $n$-го порядка и линейными преобразованиями $n$ неизвестных. Это значит, что умножая характеристики на квадратную матрицу, мы просто меняем систему координат, в которой производятся вычисления. Такое преобразования не затрагивает вычислительный алгоритм линейной регрессии.

## Алгоритм преобразования

Основываясь на приведенных выше рассуждениях, будем реализовывать следующий алгоритм шифрования персональных данных. 

**Алгоритм шифрования данных клиентов**

1.	С помощью функции np.random.normal() сгенерируем случайную обратимую квадратную матрицу В. Размерность матрицы равна количеству характеристик (в нашем случае 4).
2.	Умножим матрицу характеристик на матрицу В справа.


**Обоснование**

Утверждение 1 позволяет нам реализовать такой алгоритм действий без ущерба для качества линейной регрессии. В то же время, внешний вид данных пользователей станет нечитаемым для человека. 

Матрица шифрования должна быть обратимой по двум причинам.

Во-первых, потому что ее определитель должен быть отличен от нуля. Если определитель равен нулю, значит в матрице есть нулевая строка (столбец) или есть линейно зависимые строки. В результате умножения на такую матрицу мы получим матрицу тоже с нулевой строкой (столбцом) или с линейно зависимыми строками. Применительно к задаче машинного обучения это означает, что при таком шифровании возникнет мультиколлинеарность характеристик там, где их изначально не было. А это приведет к ухудшению качества модели.

Во-вторых, наличие обратной матрицы позволит нам в дальнейшем восстановить начальный вид данных по зашифрованному.


## Проверка алгоритма

### Создание матриц шифрования и дешифрования

In [11]:
def get_rand_matrix(dim):
    '''
    создает случайную квадратную матрицу размером dim Х dim
    '''
    det = 0
    while det == 0:
        matrix = np.random.normal(size=[dim, dim])
        det = np.linalg.det(matrix)
    return matrix

Создадим случайную квадратную матрицу размерностью 4.

In [12]:
# матрица для шифрования
B = get_rand_matrix(features.shape[1])
B

array([[ 0.49671415, -0.1382643 ,  0.64768854,  1.52302986],
       [-0.23415337, -0.23413696,  1.57921282,  0.76743473],
       [-0.46947439,  0.54256004, -0.46341769, -0.46572975],
       [ 0.24196227, -1.91328024, -1.72491783, -0.56228753]])

Cохраним обратную матрицу в отдельную переменную для последующего использования.

In [13]:
# матрица для дешифрования
B_1 = np.linalg.inv(B)
B_1

array([[-0.33124857, -1.0333483 , -2.21070923, -0.47651024],
       [ 0.12864778, -0.47781825,  0.15271687, -0.43017975],
       [-0.51338817,  0.33588835, -0.8347692 , -0.24072429],
       [ 0.99462192,  0.15079316,  1.08985145,  0.21872535]])

### Шифрование данных

Зашифруем данные клиентов и сохраним их в отдельную переменную.

In [14]:
# зашифрованные данные
features_new = features @ B
features_new.columns = features.columns
print('Размерность зашифрованных данных:', features_new.shape)
features_new.head()

Размерность зашифрованных данных: (4847, 4)


Пол       Возраст      Зарплата   Члены семьи
0 -23294.791154  26899.327002 -22921.847067 -23067.770211
1 -17850.555758  20604.598076 -17538.953455 -17662.990926
2  -9865.752553  11386.970944  -9685.974377  -9758.069218
3 -19581.515190  22616.010381 -19294.804157 -19405.939170
4 -12259.341053  14154.123038 -12050.336135 -12132.535366

Видим, что данные изменили свой внешний вид. 

Интересно, что в столбце Пол теперь не два уникальных значения. Это связано с тем, что при умножении на матрицу $В$ сюда попал результат скалярного произведения всей строки на первый столбец $В$, а значит на формирование значения оказали влияние все числа строки матрицы характеристик, что и привело к совершенно разным значениям для разных клиентов. 

Проверим обратимость шифрования данных. Для этого домножим преобразованные данные на матрицу для дешифровки.

In [15]:
# восстановленные данные
decrypted = (features_new @ B_1)
decrypted.columns = features_new.columns
decrypted.head()

Пол  Возраст  Зарплата   Члены семьи
0  1.000000e+00     41.0   49600.0  1.000000e+00
1  1.168548e-11     46.0   38000.0  1.000000e+00
2  4.429902e-12     29.0   21000.0  1.109661e-12
3  9.105370e-12     21.0   41700.0  2.000000e+00
4  1.000000e+00     28.0   26100.0  2.769852e-13

In [16]:
# округлим до целых чисел
decrypted['Пол'] = abs(round(decrypted['Пол']))
decrypted['Члены семьи'] = abs(round(decrypted['Члены семьи']))

decrypted.head()

Пол  Возраст  Зарплата  Члены семьи
0  1.0     41.0   49600.0          1.0
1  0.0     46.0   38000.0          1.0
2  0.0     29.0   21000.0          0.0
3  0.0     21.0   41700.0          2.0
4  1.0     28.0   26100.0          0.0

In [17]:
# сравним с первоначальными данными
features.head()

Пол  Возраст  Зарплата  Члены семьи
0    1     41.0   49600.0            1
1    0     46.0   38000.0            1
2    0     29.0   21000.0            0
3    0     21.0   41700.0            2
4    1     28.0   26100.0            0

Наш алгоритм шифрования позволяет восстановить первоначальный вид данных при наличии ключа дешифровки.

### Подготовка данных для обучения модели

Разобьем данные на две выборки – обучающую и тестовую в отношении 4 : 1.

In [18]:
features_train, features_test, \
features_new_train, features_new_test, \
target_train, target_test = train_test_split(
    features,
    features_new,
    target,
    test_size=.2,
    random_state=RANDOM_STATE
)

In [19]:
# проверка размерности
features_train.shape, features_test.shape

((3877, 4), (970, 4))

In [20]:
# проверка размерности
features_new_train.shape, features_new_test.shape

((3877, 4), (970, 4))

In [21]:
# проверка размерности
target_train.shape, target_test.shape

((3877,), (970,))

### Проверка на модели линейной регрессии

Напишем класс линейной регрессии.

In [22]:
class MyLinearRegression:
    
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = np.linalg.inv(X.T @ X) @ X.T @ y
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0

Обучим модель линейной регрессии на незашифрованных данных и вычислим для нее показатель метрики R2.

In [23]:
# наша модель, данные без шифрования 
model_1 = MyLinearRegression()
model_1.fit(features_train, target_train)
predictions = model_1.predict(features_test)

# метрика R2
r2_first = r2_score(target_test, predictions)
print(r2_first)

0.4160549216151218


Теперь обучим модель линейной регрессии на зашифрованных данных и сравним ее показатель метрики R2 с первым результатом.

In [24]:
# наша модель, данные зашифрованы 
model_2 = MyLinearRegression()
model_2.fit(features_new_train, target_train)
predictions_new = model_2.predict(features_new_test)

# метрика R2
r2_second  = r2_score(target_test, predictions_new)
print(r2_second)

0.41605492087348384


In [25]:
f'{r2_first - r2_second:0.10f}'

'0.0000000007'

Видим, что метрики практически совпадают. Разница появляется только в 10ом знаке после запятой.

Посмотрим на веса моделей и выясним, выполняется ли соотношение (7) из доказательства утверждения 1.

In [26]:
# веса модели без шифрования
model_1.w

array([ 1.52535832e-02,  3.64619543e-02,  1.79690961e-07, -1.41285669e-02])

In [27]:
# веса модели на зашифрованных данных
model_2.w

array([-0.03599862, -0.009382  ,  0.00781707,  0.01757968])

In [28]:
# домножим слева веса без шифрования на матрицу, обратную к В
B_1 @ model_1.w

array([-0.03599862, -0.009382  ,  0.00781708,  0.01757968])

Видим, что соотношение (7) выполняется.

### Проверка на модели из Scikit Learn

Проведем аналогичную проверку для модели линейной регрессии из Scikit Learn.

In [29]:
# модель из Scikit Learn, данные без шифрования
model_3 = LinearRegression()
model_3.fit(features_train, target_train)
predictions_3 = model_3.predict(features_test)

# метрика R2
r2_third = r2_score(target_test, predictions_3)
print(r2_third)

0.4160549216151235


In [30]:
# модель из Scikit Learn, данные зашифрованы
model_4 = LinearRegression()
model_4.fit(features_new_train, target_train)
predictions_4 = model_4.predict(features_new_test)

# метрика R2
r2_fourth = r2_score(target_test, predictions_4)
print(r2_fourth)

0.41605492161522994


In [31]:
f'{r2_third - r2_fourth:0.14f}'

'-0.00000000000011'

Метрики также совладают, разница появляется в 13ом знаке после запятой.

In [32]:
# веса модели без шифрования
model_3.coef_

array([ 1.52535832e-02,  3.64619543e-02,  1.79690961e-07, -1.41285669e-02])

In [33]:
# веса модели на зашифрованных данных
model_4.coef_

array([-0.03599862, -0.009382  ,  0.00781708,  0.01757968])

In [34]:
# домножим слева веса без шифрования на матрицу, обратную к В
np.linalg.inv(B) @ model_3.coef_

array([-0.03599862, -0.009382  ,  0.00781708,  0.01757968])

Видим, что соотношение (7) также выполняется.

### Вывод

На этапе проверки влияния алгоритма шифрования на точность моделей линейной регрессии мы работали с моделью, написанной с использованием аналитического решения нормального уравнения, и с моделью из библиотеки Scikit Learn. 

Для каждой модели провели обучение на зашифрованных и незашифрованных данных и вычислили значение материки R2. В обоих случаях значения материк совпали. Расхождение появляется после 15го знака после запятой. Скорее всего, оно связано с вычислительной погрешностью при работе с действительными числами.

Также были проверены соотношения весов моделей, полученные в ходе доказательства утверждения 1. Они также выполняются, причем для моделей из библиотеки Scikit Learn без погрешностей.


## Общий вывод

Перед нами **стоит задача зашифровать данные клиентов** страховой компании «Хоть потоп» так, чтобы по ним было сложно восстановить персональную информацию, и при этом не ухудшилось качество предсказаний модели линейной регрессии.

Нам **доступны данные 5000 клиентов**. По каждому клиенту предоставлены значения 4х числовых характеристик (пол, возраст, зарплата застрахованного и количество членов семьи). Целевой признак для модели – количество страховых выплат. Пропуски отсутствуют. В данных найдены дубликаты.

На этапе подготовки данных дубликаты удалены. Данные разбиты на характеристики и целевой признак.

**Для шифрования** персональных данных решено **матрицу признаков умножить на обратимую квадратную матрицу**, размерность которой совпадает с числом характеристик. 

Для подтверждения допустимости такого преобразования признаков мы математически **доказали** следующее **утверждение**:  

Качество модели линейной регрессии не изменится при умножении матрицы характеристик на обратимую матрицу, имеющую размерность, равную количеству характеристик, описывающих данные.

В процессе доказательства были **выведены соотношения**, которые показывают, **как соотносятся между собой векторы весов моделей обученных с применением шифрования и без него**:

$$
w = B \hat w
$$

$$
\hat w = B^{-1}  w
$$

На этапе вычислительной **проверки влияния алгоритма шифрования на точность моделей линейной регрессии** мы работали с моделью, написанной с использованием аналитического решения нормального уравнения, и с моделью из библиотеки Scikit Learn. 

Для каждой модели провели обучение на зашифрованных и незашифрованных данных и вычислили значение материки R2. В обоих случаях **значения материк совпали**. Расхождение появляется после 14го знака после запятой. 

Были проверены **соотношения весов моделей**, полученные в ходе доказательства утверждения. Они также **выполняются**, причем для моделей из библиотеки Scikit Learn без погрешностей.